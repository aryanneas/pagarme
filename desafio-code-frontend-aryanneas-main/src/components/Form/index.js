import React from 'react'
import PropTypes from 'prop-types'
import { Formik } from 'formik'

import Button from '../../components/Button'

const Form = ({
  children,
  name,
  initialValues,
  validationSchema,
  onSubmit,
  onSubmitLabel,
}) => {
  const submitFunction = async (values, helpers) => {
    const { setSubmitting } = helpers

    await onSubmit(values, helpers)

    setSubmitting(false)
  }

  return (
    <Formik
      initialValues={initialValues}
      onSubmit={submitFunction}
      validationSchema={validationSchema}
      validateOnMount
    >
      {({ handleSubmit, isValid, isSubmitting }) => (
        <form name={name} onSubmit={handleSubmit}>
          {children}

          <Button
            disabled={!isValid || isSubmitting}
            data-testid={`${name}-submit-button`}
            type="submit"
            name={onSubmitLabel}
          />
        </form>
      )}
    </Formik>
  )
}

Form.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.elementType,
    PropTypes.object,
    PropTypes.array,
    PropTypes.element,
  ]),
  name: PropTypes.string.isRequired,
  initialValues: PropTypes.object.isRequired,
  validationSchema: PropTypes.object,
  onSubmit: PropTypes.func.isRequired,
  onSubmitLabel: PropTypes.string,
}

export default Form
