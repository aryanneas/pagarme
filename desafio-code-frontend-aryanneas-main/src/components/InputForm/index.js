import React from 'react'
import PropTypes from 'prop-types'
import { useField } from 'formik'

import {
  Label,
  InputContainer,
  MaskInput,
  Input,
  ErrorContainer,
} from './style'

const InputForm = ({
  name,
  label,
  type,
  mask,
  required,
  revertMask,
  handleChange,
  ...props
}) => {
  const [
    { onChange, onBlur, ...input },
    { touched, error },
    { setValue },
  ] = useField(name)

  const onFieldChange = (e) => {
    onChange(e)
    if (handleChange) handleChange(e)
    if (mask && revertMask) setValue(revertMask(e.target.value))
  }

  const onFieldBlur = (e) => {
    onBlur(e)
    if (mask && revertMask) setValue(revertMask(e.target.value))
  }

  const InputFormLabel = ({ label, id }) => (
    <Label className="control-label" htmlFor={id}>
      {`${label}`}
    </Label>
  )

  return (
    <InputContainer>
      {label && <InputFormLabel label={label} id={name} />}

      {mask ? (
        <MaskInput
          mask={mask}
          guide={false}
          id={name}
          type={type}
          error={touched && error ? error : undefined}
          onChange={onFieldChange}
          onBlur={onFieldBlur}
          {...input}
          {...props}
          data-testid={`${name}-input-field`}
        />
      ) : (
        <Input
          id={name}
          type={type}
          error={touched && error ? error : undefined}
          onChange={onFieldChange}
          onBlur={onFieldBlur}
          {...input}
          {...props}
          data-testid={`${name}-input-field`}
        />
      )}
      {touched && <ErrorContainer>{error}</ErrorContainer>}
    </InputContainer>
  )
}

InputForm.defaultProps = {
  type: 'text',
}

InputForm.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.string,
  type: PropTypes.string,
  mask: PropTypes.oneOfType([PropTypes.array, PropTypes.func]),
  revertMask: PropTypes.func,
  required: PropTypes.bool,
  placeholder: PropTypes.string,
  handleChange: PropTypes.func,
}

export default InputForm
