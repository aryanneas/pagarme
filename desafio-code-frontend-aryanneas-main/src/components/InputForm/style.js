import styled from 'styled-components'
import MaskedInput from 'react-text-mask'

export const Input = styled.input`
  font-size: 18px;
  padding: 0 0 0 10px;
  border: 1px solid ${(props) => (props.error ? 'red' : '#8B8B92')};
  border-radius: 6px;
  width: 100%;
  height: 48px;
  margin: 10px 0;
  color: #595a63;
  &:focus {
    border: 1px solid ${(props) => (props.error ? 'red' : '#8B8B92')};
  }
`

export const Label = styled.span`
  position: relative;
  background: #fff;
  left: 20px;
  color: #8b8b92;
  top: 20px;
`

export const MaskInput = styled(MaskedInput)`
  font-size: 18px;
  padding: 0 0 0 10px;
  margin: 10px 0;
  border: 1px solid ${(props) => (props.error ? 'red' : '#8B8B92')};
  border-radius: 6px;
  width: 100%;
  height: 40px;
  &:focus {
    &::placeholder {
      color: transparent;
    }
    border: 1px solid ${(props) => (props.error ? 'red' : '#8B8B92')};
  }
`

export const InputContainer = styled.div`
  width: 100%;
  padding: 0 10px;
`

export const ErrorContainer = styled.p`
  padding: 0;
  margin: 0;
  color: #ff4040;
  font-size: 16px;
  width: 100%;
`
