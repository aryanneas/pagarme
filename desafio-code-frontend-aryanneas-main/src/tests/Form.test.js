import React from 'react'
import { render, waitForElement, cleanup } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import Form from '../components/Form'
import InputForm from '../components/InputForm'

const FORM_NAME = 'test'
const inputs = ['first', 'second']
const initialValues = {}
const handleSubmit = jest.fn()

const submitButtonIsDisabled = async (getByTestId) => {
  const currentButton = await waitForElement(() => getByTestId('test-button'))
  return currentButton.disabled
}

describe('Form component', () => {
  afterEach(cleanup)

  it('should render', () => {
    const { asFragment, getByTestId } = render(
      <Form
        name={FORM_NAME}
        onSubmit={handleSubmit}
        initialValues={initialValues}
        onSubmitLabel={FORM_NAME}
      />
    )
    expect(asFragment()).toMatchSnapshot()

    const submitButton = getByTestId('test-button')
    expect(submitButton.disabled).toBe(false)
  })

  it('should only allow submission if all fields are filled correctly', async () => {
    const { asFragment, getByTestId } = render(
      <Form
        name={FORM_NAME}
        onSubmit={handleSubmit}
        onSubmitLabel={FORM_NAME}
        initialValues={initialValues}
      >
        <InputForm required name="first" label="first input" />
        <InputForm required name="second" label="second input" />
      </Form>
    )

    expect(await submitButtonIsDisabled(getByTestId)).toBe(false)

    inputs.forEach((name) => {
      const input = getByTestId(`${name}-input-field`)
      userEvent.type(input, `hello world ${name}`)
    })
    expect(await submitButtonIsDisabled(getByTestId)).toBe(false)

    expect(asFragment()).toMatchSnapshot()
  })
})
