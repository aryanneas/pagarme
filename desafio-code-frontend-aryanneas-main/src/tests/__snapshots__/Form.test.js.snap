// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component should only allow submission if all fields are filled correctly 1`] = `
<DocumentFragment>
  <form
    name="test"
  >
    <div
      class="sc-jSgupP ForCf"
    >
      <span
        class="sc-hKgILt dxNXzf control-label"
        for="first"
      >
        first input
      </span>
      <input
        class="sc-dlfnbm kUGxZo"
        data-testid="first-input-field"
        id="first"
        name="first"
        type="text"
        value="hello world first"
      />
    </div>
    <div
      class="sc-jSgupP ForCf"
    >
      <span
        class="sc-hKgILt dxNXzf control-label"
        for="second"
      >
        second input
      </span>
      <input
        class="sc-dlfnbm kUGxZo"
        data-testid="second-input-field"
        id="second"
        name="second"
        type="text"
        value="hello world second"
      />
    </div>
    <button
      class="sc-bdfBwQ gtOdDW"
      data-testid="test-button"
    >
      test
    </button>
  </form>
</DocumentFragment>
`;

exports[`Form component should render 1`] = `
<DocumentFragment>
  <form
    name="test"
  >
    <button
      class="sc-bdfBwQ gtOdDW"
      data-testid="test-button"
    >
      test
    </button>
  </form>
</DocumentFragment>
`;
